AMLII-1701 Enabling stream-logs in a flare triggered by remote config

DDuongNguyen opened the pull request:
### What does this PR do?

This PR integrates the streamlogs functionality into the remote configuration for debug flare option. It allows support team to get visibility on their log collection process to troubleshoot missing logs using remote config. 

### Motivation

The motivation behind this change is to improve visibility and debugging capabilities for troubleshooting missing logs process. By streaming logs directly from the remote configuration client, engineers and supports can gain immediate insights into the configuration process and quickly identify and resolve issues.

### Possible Drawbacks / Trade-offs

One possible drawback of this change is the potential increase in network traffic or flare creation time due to the streaming of logs (by default it adds additional 60s). This could impact performance, especially in environments with limited network resources. However, the benefits of improved visibility and debugging capabilities are likely to outweigh this potential drawback.

### Describe how to test/QA your changes

To test these changes, follow these steps:
1. Adjust the datadog.yaml file to enable the log agent and then enable stream-logs:
```
logs_enabled: true
streaming: 
    enable_streamlogs: true
```

2. Restart the agent and check if the config is applied by running:
```
datadog-agent config get enable_streamlogs
```

As of right now since rc args are not yet available but will be by rc team, I've been testing by instead of using RC args logic, i put the streamlogs logic into the debug level mode from fleet automation. Once the logic runs cleanly and the feature is functional, i put the logic of streamlogs back behind the rc arg. 

### Possible Drawbacks / Trade-offs
Gathering streamlogs information in the flare can increase generation times. By default it is 60 seconds + network delays

DDuongNguyen committed:
add streamlog runtime setting detection to remote config flare

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
lint test

DDuongNguyen committed:
Merge branch 'remote-config-streamlogs-runtime' of github.com:DataDog/datadog-agent into remote-config-streamlogs-runtime

DDuongNguyen committed:
windows lint

DDuongNguyen committed:
Delete test/new-e2e/tests/agent-metric-logs/log-agent/macos-log/testings/testing_test.go

timothyalexandersoftware commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -297,6 +301,14 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
 		err = rc.settingsComponent.SetRuntimeSetting("log_level", mergedConfig.LogLevel, model.SourceRC)
 	}
 
+	switch streamLogsSource {
+	case model.SourceRC:
+		if *mergedConfig.EnableStreamLogs {
+			err = rc.settingsComponent.SetRuntimeSetting("enable_stream_logs", *mergedConfig.EnableStreamLogs, model.SourceRC)
+		}
+	default:
+		err = rc.settingsComponent.SetRuntimeSetting("enable_stream_logs", false, model.SourceRC)
+	}
```
Not sure I fully understand this logic. If I'm reading it right:
- if the config source for stream logs is RC, set the runtime setting to whatever is in the configuration payload from RC
- if the config source is _not_ RC, disable the runtime setting.

If that's the case, it doesn't seem like anything can ever set the runtime setting to be sourced from RC, right? My intuition is that handling for this config value should be similar to the handling of the `LogLevel` attribute, to maintain the order of precedence for runtime settings.

coignetp replied:
Something is missing here but the logic it's close. We basically need to implement the config priority here similar to `LogLevel` as Tim mentioned ; it should be something like:
* If the config is set by RC (`mergedConfig.EnableStreamLogs != nil`), it should be set accordingly
* If not but it was previously set by RC (`mergedConfig.EnableStreamLogs != nil && streamLogsSource == model.SourceRC`), revert `config.Datadog().UnsetForSource("enable_stream_logs", model.SourceRC)`
* Else, just do nothing and leave it as is

Also, I'm wondering if you should not register the `enable_stream_logs` key just like `log_level` and the other runtime settings (see [here](https://github.com/DataDog/datadog-agent/blob/201b789380d9c9faac6339d714aced6f0f4a45d8/pkg/config/settings/runtime_setting_log_level.go))

timothyalexandersoftware commented on pkg/remoteconfig/state/agent_config.go:
```diff
@@ -26,7 +26,8 @@ type AgentConfig struct {
 type ConfigContent struct {
-	LogLevel string `json:"log_level"`
+	LogLevel         string `json:"log_level"`
+	EnableStreamLogs *bool
```
It looks like this implies that there are going to be changes to the `AGENT_TASK` remote config product (and potentially the UI, to let the user toggle whether they want to enable streaming logs during flare collection?) If you haven't already, you should reach out to the Fleet Automation team at #fleet-automation, and include them in this review since they own the flare collection via RC 😄

coignetp replied:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏This needs a `json:"enable_stream_logs"` to be parsed, so we can plug it with RC

timothyalexandersoftware commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -55,6 +56,8 @@ type rcClient struct {
 	// Tasks are separated from the other products, because they must be executed once
 	taskListeners     []types.RCAgentTaskListener
 	settingsComponent settings.Component
+	Log               log.Component
+	Config            coreConfig.Component
```
I don't see these being used or set anywhere. Can they be removed?

timothyalexandersoftware commented on cmd/systray/command/command.go:
```diff
@@ -27,6 +27,8 @@ import (
 	"github.com/DataDog/datadog-agent/comp/core/config"
 	"github.com/DataDog/datadog-agent/comp/core/flare"
 	"github.com/DataDog/datadog-agent/comp/core/log/logimpl"
+	rcsetting "github.com/DataDog/datadog-agent/comp/core/settings"
```
This aliased import looks misleading. These are not RC settings. Can the alias be renamed?

timothyalexandersoftware commented on comp/core/flare/flare.go:
```diff
@@ -27,6 +29,7 @@ import (
 	"github.com/DataDog/datadog-agent/comp/core/flare/types"
 	"github.com/DataDog/datadog-agent/comp/core/log"
 	"github.com/DataDog/datadog-agent/comp/core/secrets"
+	rcsetting "github.com/DataDog/datadog-agent/comp/core/settings"
```
This aliased import looks misleading. These are not RC settings. Can the alias be renamed?

timothyalexandersoftware commented on comp/core/flare/flare.go:
```diff
@@ -44,6 +47,7 @@ type dependencies struct {
 
 	Log                   log.Component
 	Config                config.Component
+	RCsetting             rcsetting.Component
```
Same thing here, these are not remote config settings. Can the `rcsetting` fields in the `dependencies` and `flare` structs be renamed?

DDuongNguyen committed:
adjust stream log to non-cli

DDuongNguyen committed:
Merge branch 'remote-config-streamlogs-runtime' of github.com:DataDog/datadog-agent into remote-config-streamlogs-runtime

coignetp commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -297,6 +297,17 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
+	// If the config is set by RC, apply it 
+	if mergedConfig.EnableStreamLogs != nil {
+		err = rc.settingsComponent.SetRuntimeSetting("enable_stream_logs", *mergedConfig.EnableStreamLogs, model.SourceRC)
+	} else if mergedConfig.EnableStreamLogs != nil && streamLogsSource == model.SourceRC {
```
```suggestion
	} else if mergedConfig.EnableStreamLogs == nil && streamLogsSource == model.SourceRC {
```

DDuongNguyen committed:
lint

DDuongNguyen committed:
comment lint

DDuongNguyen committed:
naming lint

dustmop commented on cmd/agent/subcommands/flare/command.go:
```diff
@@ -102,11 +104,14 @@ func Commands(globalParams *command.GlobalParams) []*cobra.Command {
 					SysprobeConfigParams: sysprobeconfigimpl.NewParams(sysprobeconfigimpl.WithSysProbeConfFilePath(globalParams.SysProbeConfFilePath)),
 					LogParams:            logimpl.ForOneShot(command.LoggerName, "off", false),
 				}),
+				fx.Provide(func(c config.Component) coresetting.Params {
```
```suggestion
				fx.Supply(coresetting.Params{}),
```

dustmop commented on comp/api/api/utils/common.go:
```diff
@@ -7,11 +7,22 @@
 
+	pkgconfig "github.com/DataDog/datadog-agent/pkg/config"
+
+	"github.com/DataDog/datadog-agent/pkg/api/util"
```
Would recommend aliasing this to `apiutil` to make usages more obvious:

```suggestion
	apiutil "github.com/DataDog/datadog-agent/pkg/api/util"
```

dustmop commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -7,17 +7,30 @@
 
+// LogParams represents the parameters for streaming logs
+type LogParams struct {
+	// Output represents the output file path to write the log stream to.
```
```suggestion
	// FilePath represents the output file path to write the log stream to.
```

dustmop commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -92,3 +105,41 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+
+// ExportStreamLogs export output of stream-logs to a file
+func ExportStreamLogs(config config.Component, streamLogParams *LogParams) error {
+	ipcAddress, err := pkgconfig.GetIPCAddress()
```
Is this the best place for this functionality? All other usages of `GetIPCAddress` are in entry points, such as `cmd/.../subcommands/...` or `pkg/.../subcommands/...`, whereas this is within a utility package of the API component. Putting an IPC into a subcommand entry point makes sense because a command-line invocation of the agent can contact a running agent server process, but here it seems like the code may already be running in an agent server? I'm not convinced this makes sense from a code organization standpoint.

dustmop commented on comp/core/flare/flare.go:
```diff
@@ -155,6 +161,12 @@ func (f *flare) Send(flarePath string, caseID string, email string, source helpe
 // Create creates a new flare and returns the path to the final archive file.
 func (f *flare) Create(pdata ProfileData, ipcError error) (string, error) {
+	// Export an output of streamlog command if runtime (enable_stream_logs) is enabled so flare can attach it.
+	err := f.streamLogsIfEnabled()
```
Typically, instead of adding flare creation logic here, code will create a FlareProvider that lives in its own component, and fx value groups will tie all the flare creation code together. If you could do the same, it wouldn't be necessary to add dependencies on `core/settings` and `api/utils/stream` here, as they're not really necessary for creating the flare component.

DDuongNguyen committed:
comments

DDuongNguyen committed:
- Removes remote config streamlog logic from CLI flare command
+ Turns remote stream-logs flare into a component
+ Changes where remote config flare places dogstatsd logs

DDuongNguyen committed:
- Removes remote config streamlog logic from CLI flare command
+ Turns remote stream-logs flare into a component
+ Changes where remote config flare places dogstatsd logs

DDuongNguyen committed:
Update flare_controller.go

DDuongNguyen committed:
lint

DDuongNguyen committed:
lint

DDuongNguyen committed:
Update flare_controller.go

DDuongNguyen committed:
CI Lints

DDuongNguyen committed:
merged main and resolve conflicts

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
resolve conflicts

DDuongNguyen committed:
component lint

DDuongNguyen committed:
Merge branch 'remote-config-streamlogs-runtime' of github.com:DataDog/datadog-agent into remote-config-streamlogs-runtime

DDuongNguyen committed:
lint

DDuongNguyen committed:
add readme, add streamlog path to config files

DDuongNguyen committed:
windows lint

coignetp commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -294,6 +294,14 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
 
+	streamLogsSource := config.Datadog().GetSource("enable_stream_logs")
+
+	if mergedConfig.EnableStreamLogs != nil {
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review (if it exists :^)  -->
**👏 praise:** ‏

DDuongNguyen committed:
adding struct tag

dustmop commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +106,77 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+	if err != nil {
+		return nil, nil, fmt.Errorf("error opening file %s: %v", filePath, err)
+	}
+	bufWriter := bufio.NewWriter(f) // default 4096 bytes buffer
```
Does the comment about the buffer size affect either parts of the code here? If it's just a piece of implementation trivia I think it's okay to remove it.

dustmop commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +106,77 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+	var f *os.File
+	var bufWriter *bufio.Writer
+
+	err := CheckDirExists(streamLogParams.FilePath)
```
```suggestion
	if err := CheckDirExists(streamLogParams.FilePath); err != nil {
```

dustmop reviewed (commented):
Thanks for taking on the refactoring on this! Just a few other comments and then I think we can merge this.

vickenty commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -294,6 +294,14 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
 		err = rc.settingsComponent.SetRuntimeSetting("log_level", mergedConfig.LogLevel, model.SourceRC)
 	}
 
+	streamLogsSource := config.Datadog().GetSource("enable_stream_logs")
```
Would it make sense to unify spelling of `stream_logs` vs `streamlogs`?

Do we need to register this setting in `setup/config.go`?

Would it make sense to add these settings to the `config_template.yaml`?

vickenty commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,114 @@
+func (sl *streamlogsimpl) fillFlare(fb flaretypes.FlareBuilder) error {
+	streamlogsLogFile := sl.config.GetString("streamlogs_log_file")
+	if streamlogsLogFile == "" {
+		streamlogsLogFile = commonpath.DefaultStreamlogsLogFile
```
Would it make sense to set the default value in `setup/config.go` so we don't have to check for an empty value?

vickenty commented on comp/logs/streamlogs/mock/mock.go:
```diff
@@ -0,0 +1,19 @@
+
+// Mock returns a mock for streamlogs component.
+func Mock() streamlogs.Component {
+	// TODO: Implement the streamlogs mock
```
Is this something that you were planning to address in this PR? If not, please remove the TODO.

vickenty commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -294,6 +294,14 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
+	streamLogsSource := config.Datadog().GetSource("enable_stream_logs")
+
+	if mergedConfig.EnableStreamLogs != nil {
+		err = rc.settingsComponent.SetRuntimeSetting("enable_stream_logs", *mergedConfig.EnableStreamLogs, model.SourceRC)
```
Assigning to `err` here discards any previous error we may have received on line 293. Is this intended?

DDuongNguyen committed:
first change

DDuongNguyen committed:
refactor

DDuongNguyen committed:
Delete pkg/config/setup/go.mod

DDuongNguyen committed:
merged main

DDuongNguyen committed:
Merge branch 'remote-config-streamlogs-runtime' of github.com:DataDog/datadog-agent into remote-config-streamlogs-runtime

DDuongNguyen committed:
Add pkg/config/setup/go.mod from main branch

DDuongNguyen committed:
lint

DDuongNguyen committed:
lint

DDuongNguyen committed:
lint

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
remove unused package

DDuongNguyen committed:
move utils functions logic

DDuongNguyen committed:
spacing and lints

DDuongNguyen committed:
Merge branch 'rc-stream-logs' into remote-config-streamlogs-runtime

DDuongNguyen committed:
remove unused go.mod files

dustmop reviewed (approved):
LGTM

vickenty commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +106,72 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+		if err != nil {
+			log.Errorf("Error flushing buffer for log stream: %v", err)
+		}
+		f.Close()
```
Would it make sense to handle error from `f.Close()` here?

vickenty commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +106,72 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+
+// ExportStreamLogs export output of stream-logs to a file. Currently used for remote config stream logs
+func ExportStreamLogs(la logsAgent.Component, streamLogParams *LogParams) error {
+	var f *os.File
+	var bufWriter *bufio.Writer
```
Do we really need to declare these variables in advance?
```suggestion
```

vickenty commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	// shouldIncludeFunc ensures that only correct extension/suffix log files are collected from the streamlogs_info folder
+	// This include log roll over files eg: streamlogs.log.1 and to exclude non log files that might be present in the folder.
```
I might be missing something, what would produce this numbered file?

vickenty commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -291,14 +294,26 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
 	// Apply the new status to all configs
 	for cfgPath := range updates {
-		if err == nil {
+		if errs == nil {
```
A slice can be non-nil but still empty. Not with the current code, but checking length would be more future-proof:

```suggestion
		if len(errs) == 0 {
```

vickenty commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -291,14 +294,26 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
+		if errs == nil {
 			applyStateCallback(cfgPath, state.ApplyStatus{State: state.ApplyStateAcknowledged})
 		} else {
+			err := fmt.Errorf("error while applying remote config: %s", errs)
```
Would it make sense to use `github.com/hashicorp/go-multierror` here? We already use it for similar cases elsewhere.

DDuongNguyen replied:
I didn't know we use this. Ill adjust my code to that instead 👍🏼

vickenty commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -291,14 +294,26 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
+
+	streamLogsSource := config.Datadog().GetSource("logs_config.streaming.enable_streamlogs")
+	if mergedConfig.EnableStreamLogs {
+		if err := rc.settingsComponent.SetRuntimeSetting("enable_streamlogs", mergedConfig.EnableStreamLogs, model.SourceRC); err != nil {
```
I think you also need to register the setting here: https://github.com/datadog/datadog-agent/blob/d2e2032c39896c491805aba5e2fdfa7bf0070488/cmd/agent/subcommands/run/command.go#L456-L469

vickenty commented on pkg/config/config_template.yaml:
```diff
@@ -1019,6 +1019,25 @@ api_key:
+    ## @param enable_streamlogs - bool - optional - default: true
+    ## @env DD_LOGS_CONFIG_STREAMING_ENABLE_STREAMLOGS - bool - optional - default: true  
+    ## Enable streaming logs via remote config.
+    # enable_streamlogs: true
```
Does this mean we are always copying all logs to `streamlogs_log_file`?

DDuongNguyen replied:
Yes, we don't have a filter feature yet so we are currently copying all logs from the stream-logs command to streamlogs_log_file.

GustavoCaso replied:
Would it be worth it to start with the feature off by default and run a few tests before enabling by default? \

I'm not what is the DD way of delivering new features to customers, but I can see that we would like to test something like this before enabling by default 😄

DDuongNguyen replied:
I think i made some confusions here by defaulting it to true before. The correct default runtime setting value should be false and it should be adjusted according to how remote config set the value base on their config here:
https://github.com/DataDog/datadog-agent/pull/27026/files#diff-0728f6d5ed5d0de385153398bda70f3297aeabe049f9312c5bfcc709c4359ddeR304-R309

In this case, default is false, and if agent detects a signal from remote config that it wants to stream-logs then the agent will set the runtime to true and stream logs accordingly.

The main reason for why runtime setting is necessary is because remote config uses runtime setting as a mean to trigger actions without restarting the agent. 
Eg: they use log level run time to decide whether an info/debug flare is sent.

Additionally, I think it makes more sense that the agent would [consolidate streamlogs related configurable behaviors](https://github.com/DataDog/datadog-agent/pull/27026/files#diff-a77843a3a2d99bdc21000210a1c6a0fc99871774b33d15b6779389e8299aa794R1032-R1040) such as durations and pathing and future features such as log rotations, filter . If the agent were to receive those configurable options instead I think it would leave more room for network error.

vickenty reviewed (commented):
This change probably also merits a changelog entry.

DDuongNguyen committed:
add run time setting for EnableStreamLogs, add multierror library

GustavoCaso commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +106,70 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
 }
+
+// ExportStreamLogs export output of stream-logs to a file. Currently used for remote config stream logs
+func ExportStreamLogs(la logsAgent.Component, streamLogParams *LogParams) error {
```
I think this function `ExportStreamLogs ` and the `LogParams` should be part of the `comp/logs/streamlogs/impl/streamlogs.go`. It seems that this file is mostly providing utility function for streaming. The `ExportStreamLogs` is very specific of the flare and streamlogs component.

GustavoCaso commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -291,14 +295,26 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
+		if err := rc.settingsComponent.SetRuntimeSetting("enable_streamlogs", mergedConfig.EnableStreamLogs, model.SourceRC); err != nil {
+			errs = multierror.Append(errs, err)
+		}
+	} else if !mergedConfig.EnableStreamLogs && streamLogsSource == model.SourceRC {
+		config.Datadog().UnsetForSource("logs_config.streaming.enable_streamlogs", model.SourceRC)
 	}
```
When is this condition triggered? By reading the code is not entirely clear when stream logs would not be enabled but the source is RC. 

Could you provide some context please?

DDuongNguyen replied:
I'm not very familiar with how the settings from rc would be configured, but maybe @coignetp could clarify on this end 😅 My understanding is that we have a different config file on RC side that set/unset the value for different tasks.

coignetp replied:
From RC side, we can send a `enable_stream_logs: true/false` so we can set the runtime settings (that would be done just like the log_level setting change)
It's totally fine to also set it from different parts of the code

GustavoCaso commented on comp/logs/streamlogs/fx/fx.go:
```diff
@@ -0,0 +1,21 @@
+func Module() fxutil.Module {
+	return fxutil.Component(
+		fxutil.ProvideComponentConstructor(
+			streamlogsimpl.NewRCStreamLogFlare,
```
The convention is to call the public function `NewComponent` streamlogsimpl.NewComponent,

GustavoCaso commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,113 @@
+type Provides struct {
+	compdef.Out
+
+	Comp          streamlogs.Component
```
I don't think the component is used at all. Maybe we do not need to include it as part of the `Provides` struct

GustavoCaso reviewed (changes requested):
Great work 🎉 

Still, I think we would benefit from adding more testing around the new rcclient new logic (`comp/remote-config/rcclient/rcclientimpl/rcclient_test.go`) as well to create test for  `pkg/config/settings/runtime_setting_enable_streamlogs.go`

DDuongNguyen committed:
move export function from util to streamlogs component, adjust default value of enable_streamlogs to false, spelling

DDuongNguyen commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,113 @@
+// exportStreamLogsIfEnabled streams logs when runtime is enabled
+func (sl *streamlogsimpl) exportStreamLogsIfEnabled(logAgent logsAgent.Component, streamlogsLogFilePath string) error {
+	// If the streamlog runtime setting is set, start streaming log to default file
+	enableStreamLog, err := sl.coresetting.GetRuntimeSetting("enable_stream_logs")
```
ah, this was a typo. Thanks for catching this!

DDuongNguyen committed:
add runtime testing for rcclient_test.go, remove dead code

GustavoCaso commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -34,34 +35,88 @@ type Requires struct {
+	}
+	defer func() {
+		if err = bufWriter.Flush(); err != nil {
+			log.Errorf("Error flushing buffer for log stream: %v", err)
```
I think makes more sense to use the `logsAgent` to log things rather than using the `pkg/util/log`

GustavoCaso commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -34,34 +35,88 @@ type Requires struct {
+	done := make(chan struct{})
+	defer close(done)
+
+	logChan := messageReceiver.Filter(&filters, done)
```
Since you are passing an empty filter pointer, we can use `nil` iin this case as it would have the same behavior.

From `pkg/logs/diagnostic/message_receiver.go`
```go
func shouldHandleMessage(m *messagePair, filters *Filters) bool {
	if filters == nil {
		return true
	}

	shouldHandle := true

	if filters.Name != "" {
		shouldHandle = shouldHandle && m.msg.Origin != nil && m.msg.Origin.LogSource.Name == filters.Name
	}

	if filters.Type != "" {
		shouldHandle = shouldHandle && ((m.msg.Origin != nil && m.msg.Origin.LogSource.Config.Type == filters.Type) || m.eventType == filters.Type)
	}

	if filters.Source != "" {
		shouldHandle = shouldHandle && m.msg.Origin != nil && filters.Source == m.msg.Origin.Source()
	}

	if filters.Service != "" {
		shouldHandle = shouldHandle && m.msg.Origin != nil && filters.Service == m.msg.Origin.Service()
	}

	return shouldHandle
}
```

GustavoCaso commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -70,11 +125,11 @@ func (sl *streamlogsimpl) exportStreamLogsIfEnabled(logAgent logsAgent.Component
 				Duration: 60 * time.Second, // Default duration is 60 seconds
 			}
-			if err := stream.ExportStreamLogs(logAgent, &streamLogParams); err != nil {
+			if err := exportStreamLogs(logsAgent, &streamLogParams); err != nil {
```
We can add this function as part of the `streamlogsimpl`

`sl.exportStreamLogs(&streamLogParams)` no need to pass the `logsAgent` around 😄

DDuongNguyen committed:
replace logger, remove emptry filter for nil

DDuongNguyen committed:
remove tests for future PR

GustavoCaso commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,162 @@
+func (sl *streamlogsimpl) fillFlare(fb flaretypes.FlareBuilder) error {
+	streamlogsLogFile := sl.config.GetString("logs_config.streaming.streamlogs_log_file")
+
+	la, ok := sl.logsAgent.Get()
+	if !ok {
+		return fmt.Errorf("log agent not found, unable to export stream logs")
+	}
```
I might sure, but I'm thinking that we could move this check the the initialization of the component `NewComponent`

If the logs agent is not present we return an empty flare provider. That way we avoid every time we request a flare with logs disable to hit this case. 

```go
// streamlog is a type that contains information needed to insert into a flare from the streamlog process.
type streamlogsimpl struct {
	logsAgent   logsAgent.Component
	logger      logger.Component
	config      config.Component
	coresetting coresetting.Component
}

func NewComponent(reqs Requires) (Provides, error) {
	la, ok := reqs.logsAgent.Get()
	if !ok {
		reqs.logger.Debug("logs agent is not enabled. unable to export stream logs in any flare")
		return Provides{}, nil
	}
	
	sl := &streamlogsimpl{
		logsAgent:   la,
		logger:      reqs.Logger,
		config:      reqs.Config,
		coresetting: reqs.CoreSetting,
	}

	provides := Provides{
		FlareProvider: flaretypes.NewProvider(sl.fillFlare),
	}
	return provides, nil
}
```

I modify the `streamlogsimpl` to remove the optional LogsAgent component.

DDuongNguyen committed:
-add unit test for new runtime, add log agent into streamlogs component when started

DDuongNguyen committed:
add constances descriptions for lints

DDuongNguyen committed:
lint

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
change config package name

GustavoCaso reviewed (approved):
Amazing work 🎉

vickenty commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,163 @@
+// exportStreamLogs export output of stream-logs to a file. Currently used for remote config stream logs
+func exportStreamLogs(la logsAgent.Component, logger logger.Component, streamLogParams *LogParams) error {
+	if err := stream.EnsureDirExists(streamLogParams.FilePath); err != nil {
+		return fmt.Errorf("error creating directory for file %s: %v", streamLogParams.FilePath, err)
```
```suggestion
		return fmt.Errorf("error creating directory for file %q: %w", streamLogParams.FilePath, err)
```

Using `%q` for filename would quote the string and escape any unusual characters, so it would be easier to catch errors due to accidental whitespace or control characters in the name.

`%w` would allow downstream code to inspect the underlying error (see https://pkg.go.dev/fmt#Errorf)

vickenty commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,163 @@
+	messageReceiver := la.GetMessageReceiver()
+
+	if !messageReceiver.SetEnabled(true) {
+		return fmt.Errorf("unable to enable message receiver, another client is already streaming logs")
```
```suggestion
		return errors.New("unable to enable message receiver, another client is already streaming logs")
```

vickenty commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,163 @@
+	done := make(chan struct{})
+	defer close(done)
+
+	logChan := messageReceiver.Filter(nil, done)
```
This isn't probably a bug in this code, but it seems this pattern may dead-lock the producer, if the timer fires as the `logChan` is getting full, the producer may get blocked [here](https://github.com/DataDog/datadog-agent/blob/c079441bf0d6b2fc6eeaabb95d9fae69a9da4e34/pkg/logs/diagnostic/message_receiver.go#L119) and never check the `done` channel.

vickenty commented on pkg/config/config_template.yaml:
```diff
@@ -1026,6 +1026,24 @@ api_key:
+
+    ## @param enable_streamlogs - bool - optional - default: false
+    ## @env DD_LOGS_CONFIG_STREAMING_ENABLE_STREAMLOGS - bool - optional - default: false  
+    ## Enable streaming logs via remote config.
```
I probably misunderstand something, but I think this description is inaccurate. If I understand the code here correctly, changing this setting would affect whether flares get a snippet of the logs, but doesn't affect the ability of remote config to also enable that.

vickenty commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +96,23 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+
+// OpenFileForWriting opens a file for writing
+func OpenFileForWriting(filePath string) (*os.File, *bufio.Writer, error) {
+	log.Infof("opening file %s for writing", filePath)
```
Would it make sense to add some context to this message, like why we are opening the file? Consider that the user will see the line in the logs out of context of the surrounding code.

vickenty commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +96,23 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+// OpenFileForWriting opens a file for writing
+func OpenFileForWriting(filePath string) (*os.File, *bufio.Writer, error) {
+	log.Infof("opening file %s for writing", filePath)
+	f, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```
Would it make sense to truncate the file here or clean the streaming file in some other way? Otherwise subsequent flares on the same host would keep getting bigger and bigger, possibly including outdated information.

vickenty commented on comp/core/flare/providers.go:
```diff
@@ -44,9 +44,9 @@ func (f *flare) collectLogsFiles(fb types.FlareBuilder) error {
-	fb.CopyDirToWithoutScrubbing(filepath.Dir(dogstatsdLogFile), "logs", shouldIncludeFunc) //nolint:errcheck
+	fb.CopyDirToWithoutScrubbing(filepath.Dir(logFile), "logs", shouldIncludeFunc)                         //nolint:errcheck
+	fb.CopyDirToWithoutScrubbing(filepath.Dir(jmxLogFile), "logs", shouldIncludeFunc)                      //nolint:errcheck
+	fb.CopyDirToWithoutScrubbing(filepath.Dir(dogstatsdLogFile), "logs/dogstatsd_info", shouldIncludeFunc) //nolint:errcheck
```
Why the path is changed for dogstatsd logs here?

DDuongNguyen replied:
The original code output dogstatsdlogfile into the same path as agent.log whereas it should be inside dogstatsd_info folder instead. So this is just a small fix to a remote config flare's log files placement

vickenty commented on pkg/config/settings/runtime_setting_enable_streamlogs.go:
```diff
@@ -0,0 +1,80 @@
+	StreamLogRunTimeSettingName = "enable_streamlogs"
+
+	// StreamLogRuntimeDescription is the description of the runtime setting.
+	StreamLogRuntimeDescription = "Enable/disable remote config streamlogs at runtime. Possible values: true, false"
```
Again, the description is misleading. The setting is controlled by remote config, but itself it enables inclusion of logs in the flare.

Not that there is anything wrong having a runtime setting for this, is it really necessary? RC can call `config.Set` directly, the only benefit this adds is ability to do `agent config set enable_streamlogs true`, which seems redundant given that the setting doesn't affect agent runtime unless flare is running, and an option to add logs to a flare already exists.

vickenty commented on pkg/config/settings/runtime_setting_enable_streamlogs.go:
```diff
@@ -0,0 +1,80 @@
+
+// EnableStreamLogsRuntimeSetting wraps operations to enable or disable remote config stream logs at runtime.
+type EnableStreamLogsRuntimeSetting struct {
+	enabled bool
```
This field appears to be unused?

DDuongNguyen committed:
remove enable_streamlogs runtime usage in terminal, deadlock proof the streamlog channel, change config description, making sure log file is truncated when new entry is written

add time.afterfunc

DDuongNguyen committed:
lint

vickenty commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,166 @@
+// exportStreamLogsIfEnabled streams logs when runtime is enabled
+func (sl *streamlogsimpl) exportStreamLogsIfEnabled(logsAgent logsAgent.Component, streamlogsLogFilePath string) error {
+	// If the streamlog runtime setting is set, start streaming log to default file
+	enableStreamLog, err := sl.coresetting.GetRuntimeSetting("enable_streamlogs")
```
You removed the runtime setting, but it is still being referred to here, and in a few other places, how does it work?

vickenty commented on pkg/config/config_template.yaml:
```diff
@@ -1026,6 +1026,24 @@ api_key:
+
+    ## @param streamlogs_log_file - string - optional
+    ## @env DD_LOGS_CONFIG_STREAMING_STREAMLOGS_LOG_FILE - string - optional
+    ## Path to the file containing the streamlogs log file.
```
> file containing the [...] file

DDuongNguyen committed:
switch out runtime setting with agent config, update darwin log path to the correct one, introduce wait group to stream log so logs dont get cut off

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
pkg lint

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
add flare time out and fix race condition for log channel

DDuongNguyen committed:
add default stream log duration for flare arg

DDuongNguyen committed:
lint

ddrthall commented on comp/remote-config/rcclient/rcclientimpl/rcclient.go:
```diff
@@ -311,14 +317,26 @@ func (rc rcClient) agentConfigUpdateCallback(updates map[string]state.RawConfig,
+		}
+	}
+
+	streamLogsSource := pkgconfigsetup.Datadog().GetSource("logs_config.streaming.enable_streamlogs")
```
With the new flareargs functionality we don't need to rely on enabling stream logs via a config entry and can instead directly be passed a TaskArg in flare.go:onAgentTaskEvent

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,197 @@
+
+	timer := time.NewTimer(streamLogParams.Duration)
+	defer timer.Stop()
+	var wg sync.WaitGroup
```
Unless I've missed something, I don't think this wait group is doing useful work. Having bufwriter threads being spun off of the main loop has out-of-order concerns for the logs, what problem is it solving?

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,197 @@
+			wg.Add(1)
+			go func(log string) {
+				defer wg.Done()
+				_, err := bufWriter.WriteString(log + "\n")
```
Even if the wait group + go routine is valid, bufWriter.WriteString is unfortunately not thread safe and would require additional mutex protections to be utilized here.

DDuongNguyen replied:
I made some changes to the logic of the process. I added a few items:

1. A buffered channel to decouple the producer from the consumer, preventing the producer from stalling on I/O. 

2. A dedicated consumer goroutine that batches and writes logs, using a timer to flush periodically. 

3. On shutdown, channels are closed and a WaitGroup to ensures all logs are written to avoid deadlocks and blocking.

Let me know if this could potentially help with the issue!

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+// DefaultStreamLogsDuration is 60 seconds
+const DefaultStreamLogsDuration = 60 * time.Second
```
This could be a good candidate for an undocumented config option rather than enabled_streamlogs, as it's consistent from run-to-run but some systems may want a different default

DDuongNguyen committed:
remove config for enable streamlogs, instead use duration enabled by rc flare arg. adjust streamlogs to prevent blocking and deadlock

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
Merge branch 'main' into remote-config-streamlogs-runtime

DDuongNguyen committed:
Merge branch 'remote-config-streamlogs-runtime' of github.com:DataDog/datadog-agent into remote-config-streamlogs-runtime

DDuongNguyen committed:
lint

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,234 @@
+// Requires defines the dependencies for the streamlogs component
+type Requires struct {
+	compdef.In
+	LogsAgent   optional.Option[logsAgent.Component]
```
Expected type is "option.Option" rather than "optional.Option"

ddrthall commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +96,23 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
+// EnsureDirExists checks if the directory for the given path exists, if not then create it.
+func EnsureDirExists(path string) error {
+	dir := filepath.Dir(path)
+	if err := os.MkdirAll(filepath.Dir(dir), 0755); err != nil {
```
Double "Dir"-ing

DDuongNguyen committed:
merge main, switch streamlogs blocking logic

ddrthall commented on pkg/remoteconfig/state/agent_config.go:
```diff
@@ -26,7 +26,8 @@ type AgentConfig struct {
 type ConfigContent struct {
-	LogLevel string `json:"log_level"`
+	LogLevel         string `json:"log_level"`
+	EnableStreamLogs bool   `json:"enable_stream_logs"`
```
This config component should be safe to remove

ddrthall commented on go.mod:
```diff
@@ -581,6 +581,7 @@ require (
 )
 
 require (
+	github.com/DataDog/datadog-agent/pkg/util/optional v0.59.0
```
This dependency seems safe to remove

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+// writeBatch writes a batch of logs to the buffer writer so that it can be flushed to the file.
+func writeBatch(bufWriter *bufio.Writer, batch []string, logger logger.Component) {
```
writeBatch looks safe to remove post refactor

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,194 @@
+	done := make(chan struct{})
+	logChan := messageReceiver.Filter(nil, done)
+
+	timer := time.NewTimer(streamLogParams.Duration)
```
timer looks safe to remove post refactor

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,194 @@
+		if !ok {
+			break
+		}
+		if _, err := bufWriter.WriteString(log + "\n"); err != nil {
```
Do we need the explicit newline here? Not sure if the log comes with one already embedded.

ddrthall commented on comp/core/flare/flare.go:
```diff
@@ -120,6 +120,15 @@ func (f *flare) onAgentTaskEvent(taskType rcclienttypes.TaskType, task rcclientt
 	// flareArgs.ProfileDuration = f.config.GetDuration("flare.rc_profiling_runtime") will be utilized here on completion
 	// of AMLII-2127
 
+	streamlogs, found := task.Config.TaskArgs["enable_streamlogs"]
```
Nit: Some quick unit tests can be rigged up to be reasonably sure we're parsing this correctly moving forward, given how difficult it is to test this manually

DDuongNguyen committed:
Add unit test, refactor codes, merged main.

DDuongNguyen committed:
Component lint

DDuongNguyen committed:
component lint

ddrthall commented on comp/core/flare/flare.go:
```diff
@@ -133,7 +133,7 @@ func (f *flare) onAgentTaskEvent(taskType rcclienttypes.TaskType, task rcclientt
 
 	streamlogs, found := task.Config.TaskArgs["enable_streamlogs"]
-	if !found {
+	if !found || streamlogs == "false" {
```
The nit-iest nit that was ever a nit: A streamlogs value being found but false makes the debug log message slightly incorrect. If there's nothing else you need to change feel free to ignore.

pgimalac commented on comp/README.md:
```diff
@@ -308,6 +308,12 @@ read), and tailing mode for each log file.
 
+### [comp/logs/streamlogs](https://pkg.go.dev/github.com/DataDog/datadog-agent/comp/logs/streamlogs)
+
+*Datadog Team*: agent-metrics-logs
```
Hum that's outdated I guess ?

DDuongNguyen replied:
I tried to put agent-log-pipeline but the CI LINT for component lint kept making me put agent-metric-logs instead. I guess its because this PR started back when it was still under AML

pgimalac replied:
You must update the comment in `comp/logs/streamlogs/def/component.go`, I added a suggestion

pgimalac commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +96,23 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
 		}
 	}
 }
+
+// OpenFileForWriting opens a file for writing
+func OpenFileForWriting(filePath string) (*os.File, *bufio.Writer, error) {
+	log.Infof("Opening file %s for writing logs. This file will be used to store streamlog output.", filePath)
+	f, err := os.OpenFile(filePath, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error opening file %s: %v", filePath, err)
+	}
+	bufWriter := bufio.NewWriter(f)
+	return f, bufWriter, nil
+}
+
+// EnsureDirExists checks if the directory for the given path exists, if not then create it.
+func EnsureDirExists(path string) error {
+	dir := filepath.Dir(path)
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return err
+	}
+	return nil
+}
```
Why is this defined in this package ?

DDuongNguyen replied:
These functions are used specfically forh streamlogs agent CLI command, and the new streamlogs component that have the functionality of streamlogs. 

Since those 2 are streamlogs and the package is used for streaming, i put it there.

I could move it to another utils folder if needed. Maybe something like utils/fileutils/... ?

pgimalac replied:
I meant that the package (and the function which are currently in it, eg. `GetStreamFunc`) are generic streaming helpers, not specific to logs (I think ?), so it seems weird to have those exported functions which are not about streaming being there

> I could move it to another utils folder if needed. Maybe something like utils/fileutils/... ?

Yeah I think using one of the subpackages of `pkg/util/` (or creating a new one even) would be best 👍

pgimalac commented on comp/api/api/utils/stream/stream.go:
```diff
@@ -93,3 +96,23 @@ func GetStreamFunc(messageReceiverFunc func() MessageReceiver, streamType, agent
 }
+
+// OpenFileForWriting opens a file for writing
+func OpenFileForWriting(filePath string) (*os.File, *bufio.Writer, error) {
+	log.Infof("Opening file %s for writing logs. This file will be used to store streamlog output.", filePath)
```
This log line seems too specific to logs streaming, you should probably log that before the call (the function is very generic)

ddrthall commented on comp/logs/streamlogs/impl/streamlogs.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	if err := fb.CopyDirTo(filepath.Dir(streamlogsLogFile), "logs/streamlogs_info", shouldIncludeFunc); err != nil {
+		return fmt.Errorf("failed to copy logs to flare: %w", err)
```
Do we hit this error case when profiling is disabled via a zero duration?

ddrthall reviewed (approved):
Logic seems solid to me, very nice work heaving this thing over the line.

pgimalac commented on comp/logs/streamlogs/def/component.go:
```diff
@@ -0,0 +1,12 @@
+// Package streamlogs is metadata provider for streamlogs
+package streamlogs
+
+// team: agent-metrics-logs
```
```suggestion
// team: agent-log-pipelines
```

DDuongNguyen committed:
refactor utility functions, adjust team to alp

DDuongNguyen commented:
/merge
